wagtail/admin/views/workflows.py

```diff
@@ -392,24 +398,62 @@ def delete_action(self):
         self.object.deactivate(user=self.request.user)
 
 
-def usage(request, pk):
-    workflow = get_object_or_404(Workflow, id=pk)
+class WorkflowUsageView(PageListingMixin, PermissionCheckedMixin, BaseListingView):
+    pk_url_kwarg = "pk"
+    index_url_name = "wagtailadmin_workflows:usage"
+    index_results_url_name = "wagtailadmin_workflows:usage_results"
+    paginate_by = 20
+    header_icon = "tasks"
+    page_title = _("Usage")
 
-    editable_pages = page_permission_policy.instances_user_has_permission_for(
-        request.user, "change"
-    )
-    pages = workflow.all_pages() & editable_pages
-    paginator = Paginator(pages, per_page=10)
-    pages = paginator.get_page(request.GET.get("p"))
+    def dispatch(self, request, *args, **kwargs):
+        self.object = self.get_object()
 
-    return render(
-        request,
-        "wagtailadmin/workflows/usage.html",
-        {
-            "workflow": workflow,
-            "used_by": pages,
-        },
-    )
+        # We set page_permission_policy as the main permission policy for this view
+        # via PageListingMixin, since this view is about listing pages that are using
```

Comment:
This irked me a bit as well!

I've added a few commits to hopefully improve this:

- 3457993: Remove `_` prefix from `PageListingMixin._annotate_queryset()`
  - This also changes `self.permission_policy` to `page_permission_policy` so it doesn't need to assume `self.permission_policy` is indeed a page permission policy before it can use `explorable_instances()`.
- 30832e9: Remove `permission_policy` and `any_permission_required` from `PageListingMixin`, and require subclasses to define them themselves.
- a9100c2: Remove the `.filter()` with `.explorable_instances()` from `PageListingMixin.annotate_queryset()`.
  - Seems a bit misleading that `.annotate_queryset()` would do a `.filter()`, so I moved this to each view classes instead.
  - I added the logic to pages `IndexView` and `ExplorableIndexView`'s `get_base_queryset()`. I think there _shouldn't_ be any change of behaviour to these views before vs after this PR. The only difference is that now they both do the filter in their own `get_base_queryset()` instead of relying on it from `_annotate_queryset()`.
  - I noticed that the `SearchView` and `ContentTypeUseView` currently don't seem to have any permission-related filtering 😨:
    - https://github.com/wagtail/wagtail/blob/stable/6.2.x/wagtail/admin/views/pages/search.py#L133-L160
    - https://github.com/wagtail/wagtail/blob/stable/6.2.x/wagtail/admin/views/pages/usage.py#L59-L60
    I've added the same `explorable_instances()` filter to both views now.
  - The `WorkflowUsageView` already use `instances_user_has_permission_for()`: https://github.com/wagtail/wagtail/blob/stable/6.2.x/wagtail/admin/views/workflows.py#L396-L402 so I decided to leave it as-is.

On that last part, it makes me wonder: for the "flat" listings (i.e. anything other than `ExplorableIndexView`), do we need the `explorable_instances()` filter, instead of just using `instances_user_has_any_permission_for()`? Sounds like we should just use the latter.

One particular blocker that prevented me from using it is that the `PagePermissionPolicy` currently works with the base `Page` model directly, instead of properly following the `model` vs `auth_model` distinction from the permission policy interface. This means for page listings of a specific type, `instances_user_has_any_permission_for()` will give you base `Page` instances, and you need to use `.specific()` – which seems to lose any `.select_related()` annotations in the process.

I'd like to rework `PagePermissionPolicy` to follow the `model` and `auth_model` distinction (as a separate PR), but I'm a bit afraid how that might affect wagtail/rfcs#102.

Reply:
> On that last part, it makes me wonder: for the "flat" listings (i.e. anything other than `ExplorableIndexView`), do we need the `explorable_instances()` filter, instead of just using `instances_user_has_any_permission_for()`? Sounds like we should just use the latter.

I think you're right - "explorablility" as a concept is only really needed to handle the fairly obscure case of a user having permissions over multiple distinct subtrees and needing to navigate between them. For flat listings, `instances_user_has_any_permission_for` is sufficient, and more correct. Not a blocker for this PR, though.